File: Assets/Scripts/UI/Art/MenuPack.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace LastCall.UI
{
    /// <summary>
    /// THE AUTHOR'S BUTTON PACK (2026-09-15, the author: "Butonlar içinde bu dosya yolundaki butonları kullan" —
    /// Desktop/konsept art/buton, brought in by Tools/menu_pack.py). Ten 80x96 sheets of thirty 16x16 icon buttons —
    /// grey, orange and green, each resting, lit (a brighter glyph) and pressed (the frame a pixel lower, its shadow
    /// a row thinner). The pause menu and the settings window are built from them, shown at exactly 2x:
    ///   · an ICON KEY is one cell, whole (<see cref="Cell"/>): the meters' - and +, the player's keys, the sound switch
    ///   · a WORDED KEY stands on a cell with its glyph painted out, 9-sliced at 4 (<see cref="Blank"/>), and carries
    ///     one of the pack's glyphs as a white mask at its left (<see cref="Glyph"/>), tinted in the pack's own inks
    /// The pack keeps its own palette — the one written exception to GDD 16's palette-token rule, on the author's
    /// word, the way the flags and the bottles keep theirs. Everything drawn AROUND a key stays in the palette.
    /// Sprites are cut at run time from the sheets and cached the way ChromeArt caches (a dead sprite is not a hit).
    /// </summary>
    public static class MenuPack
    {
        public enum Tone { Grey, Orange, Green }
        public enum State { Rest, Lit, Pressed }

        /// <summary>The pack's cells in sheet order, five a row — and, in a seventh row on the mask sheet only,
        /// the two the player needs that the pack lacks (the song before and after, Tools/menu_pack.py).</summary>
        private static readonly string[] Icons =
        {
            "exit", "tent", "play", "home", "stats",
            "back", "pause", "menu", "music", "gamepad",
            "expand", "sound_on", "sound_off", "cog", "save",
            "restart", "close", "trophy", "mail", "heart_line",
            "check", "trash", "chevron_down", "chevron_up", "heart",
            "plus", "minus", "lock", "unlock", "info",
            "prev", "next",
        };
        private const int CellPx = 16, Cols = 5, PackCells = 30;

        private static readonly Dictionary<string, Sprite> Cache = new Dictionary<string, Sprite>();
        private static readonly Dictionary<string, Texture2D> Sheets = new Dictionary<string, Texture2D>();

        private static Texture2D Sheet(string name)
        {
            if (Sheets.TryGetValue(name, out var t) && t != null) return t;
            t = Resources.Load<Texture2D>("Menu/" + name);
            if (t != null) Sheets[name] = t; else Sheets.Remove(name);   // a miss is never cached
            return t;
        }

        /// <summary>Which sheet carries a tone in a state. Green has no lit-and-up sheet, so its lit state is its
        /// resting one (the hover lift still says "under the pointer").</summary>
        private static string SheetName(Tone tone, State state)
        {
            string t = tone.ToString().ToLowerInvariant();
            switch (state)
            {
                case State.Lit: return tone == Tone.Green ? t + "_normal" : t + "_light";
                case State.Pressed: return t + "_light_pressed";
                default: return t + "_normal";
            }
        }

        private static Sprite Cut(Texture2D tex, string key, Rect rect, Vector4 border)
        {
            var s = Sprite.Create(tex, rect, new Vector2(0.5f, 0.5f), 100f, 0, SpriteMeshType.FullRect, border);
            s.name = key;
            return Cache[key] = s;
        }

        /// <summary>One icon button, whole — 16x16, to be shown at 2x. Null for a name the pack has no cell for.</summary>
        public static Sprite Cell(Tone tone, State state, string icon)
        {
            string key = "pack:" + tone + ":" + state + ":" + icon;
            if (Cache.TryGetValue(key, out var got) && got != null) return got;
            int i = Array.IndexOf(Icons, icon);
            if (i < 0 || i >= PackCells) return null;
            var tex = Sheet("pack_" + SheetName(tone, state));
            if (tex == null) return null;
            int r = i / Cols, c = i % Cols;   // rows count from the top of the PNG; a texture's rows from its foot
            return Cut(tex, key, new Rect(c * CellPx, tex.height - (r + 1) * CellPx, CellPx, CellPx), Vector4.zero);
        }

        /// <summary>The plate a worded key stands on: a cell with no glyph, 9-sliced at 4 so its cut corners, its
        /// highlight and its shadow stay the pack's while the middle takes the word's width.</summary>
        public static Sprite Blank(Tone tone, bool pressed)
        {
            string key = "pack:blank:" + tone + (pressed ? ":down" : ":up");
            if (Cache.TryGetValue(key, out var got) && got != null) return got;
            var tex = Sheet("pack_" + tone.ToString().ToLowerInvariant() + (pressed ? "_blank_pressed" : "_blank"));
            if (tex == null) return null;
            return Cut(tex, key, new Rect(0, 0, tex.width, tex.height), new Vector4(4, 4, 4, 4));
        }

        /// <summary>A glyph as a white mask (its second tone at half alpha), 16x16, for the caller to tint with
        /// <see cref="Ink"/>; null for an unknown name.</summary>
        public static Sprite Glyph(string icon)
        {
            string key = "pack:glyph:" + icon;
            if (Cache.TryGetValue(key, out var got) && got != null) return got;
            int i = Array.IndexOf(Icons, icon);
            var tex = Sheet("pack_glyphs");
            if (i < 0 || tex == null) return null;
            int r = i / Cols, c = i % Cols;
            return Cut(tex, key, new Rect(c * CellPx, tex.height - (r + 1) * CellPx, CellPx, CellPx), Vector4.zero);
        }

        /// <summary>
        /// THE PACK'S PLATE IN THE GAME'S OWN COLOURS (2026-09-17, the author's seventh list: "Built sahnesinde
        /// butonların rengi oyuna uygun değil butonların rengini bizim oyun tasarımımıza uygun renk paletinden
        /// seç"). The pack keeps its palette in the pause menu, where it is the whole look; on the BENCH it now
        /// stands beside a counter, a plaque and two gauges that are all cut from GDD 16's ramps, and the pack's
        /// own blue-grey and its poster orange were the only things on the screen that came from somewhere else.
        ///
        /// Not a tint: a tint multiplies, so an amber key over a mid-grey plate comes out mud. The plate is
        /// REPAINTED the way the counter's drawing is (CounterFinish.Repaint): each of the drawing's own colours
        /// is placed on the target ramp by its brightness, so the pack's bevel, its lit top edge and the row of
        /// shadow under it all survive - the key is the author's key, in the game's colours.
        /// </summary>
        public static Sprite Plate(string name, Color[] ramp, bool pressed)
        {
            string key = "pack:plate:" + name + (pressed ? ":down" : ":up");
            if (Cache.TryGetValue(key, out var got) && got != null) return got;
            var src = Blank(Tone.Grey, pressed);
            if (src == null || ramp == null || ramp.Length == 0) return src;
            var tex = src.texture;
            var px = tex.GetPixels32();

            // the drawing's own range, so the darkest pixel lands on the ramp's foot and the lightest on its head
            float lo = float.MaxValue, hi = float.MinValue;
            foreach (var c in px)
            {
                if (c.a == 0) continue;
                float l = Lum(c);
                if (l < lo) lo = l;
                if (l > hi) hi = l;
            }
            var map = new Dictionary<int, Color32>();
            for (int i = 0; i < px.Length; i++)
            {
                var c = px[i];
                if (c.a == 0) continue;
                int k = (c.r << 16) | (c.g << 8) | c.b;
                if (!map.TryGetValue(k, out var to))
                {
                    float t = hi <= lo ? 0.5f : (Lum(c) - lo) / (hi - lo);
                    to = Step(ramp, t);
                    map[k] = to;
                }
                px[i] = new Color32(to.r, to.g, to.b, c.a);
            }
            var copy = new Texture2D(tex.width, tex.height, TextureFormat.RGBA32, false)
            { filterMode = tex.filterMode, wrapMode = tex.wrapMode, name = tex.name + "_" + name };
            copy.SetPixels32(px);
            copy.Apply(false, false);
            var s = Sprite.Create(copy, src.rect, new Vector2(0.5f, 0.5f), src.pixelsPerUnit, 0,
                                  SpriteMeshType.FullRect, src.border);
            s.name = key;
            return Cache[key] = s;
        }

        private static float Lum(Color32 c) => 0.299f * c.r + 0.587f * c.g + 0.114f * c.b;

        /// <summary>A ramp read at 0..1, between its steps — CounterFinish's Ramp, for a Color[] rather than a
        /// Color32[].</summary>
        private static Color32 Step(Color[] ramp, float t)
        {
            float f = Mathf.Clamp01(t) * (ramp.Length - 1);
            int i = Mathf.Clamp(Mathf.FloorToInt(f), 0, ramp.Length - 2);
            return Color32.Lerp(ramp[i], ramp[i + 1], f - i);
        }

        /// <summary>The word's ink on a repainted plate: the palette's own answer for reading on that fill.</summary>
        public static Color WordOn(Color[] ramp) => UITheme.InkOn(Step(ramp, 0.62f));

        // ── THE PACK IN THE GAME'S OWN COLOURS (2026-09-22, the author: "ESC, Menu, butonların hepsi oyunun sanat
        // tasarımındaki renk paletlerine uygun seçilmeli; seçerken UI renklerini kullanılan paleti göz önünde
        // bulundur") ────────────────────────────────────────────────────────────────────────────────────────────
        //
        // The pack ships in its own factory colours - a poster orange, a leaf green, a blue-grey - and they were the
        // only things on the ESC menu that came from outside GDD 14's 55. The bench keys have been repainted onto
        // the palette's ramps since 2026-09-17 (KeyRamp); this is the same repaint, offered to every key the pack
        // draws, so one rule covers the menu, the settings and the bench alike:
        //
        //   GREY   -> NIGHT, the chrome ramp: what a key is when it is not asking for anything.
        //   ORANGE -> AMBER, the one accent (GDD 16 §3: PrimaryAction, one to a screen).
        //   GREEN  -> LIME, the yes: a thing that is on, a setting that is taken.
        //   HOVER  -> CLUB BLUE, the same blue the back bar answers a pointer with. Nothing else in the game
        //            is blue, so a blue plate means exactly one thing: the pointer is on this.
        public static Color[] Ramp(Tone tone) =>
            tone == Tone.Orange ? UITheme.Amber
          : tone == Tone.Green ? UITheme.Lime
          : NightKeys;

        /// <summary>Night from its second step up: the whole ramp put a key's face on Night[2] and it read as a hole
        /// cut in the page (the bench measured this on 2026-09-17).</summary>
        private static readonly Color[] NightKeys =
            { UITheme.Night[1], UITheme.Night[2], UITheme.Night[3], UITheme.Night[4] };

        /// <summary>The blue a key wears under the pointer.</summary>
        private static readonly Color[] HoverKeys =
            { UITheme.ClubBlue[1], UITheme.ClubBlue[2], UITheme.ClubBlue[3], UITheme.ClubBlue[4] };

        /// <summary>The pack's plate for this tone, cut from the palette: resting, or pressed.</summary>
        public static Sprite Paletted(Tone tone, bool pressed) =>
            Plate(tone.ToString().ToLowerInvariant(), Ramp(tone), pressed);

        /// <summary>...and the one it wears under the pointer, in the game's blue.</summary>
        public static Sprite Hovered() => Plate("hover_blue", HoverKeys, false);

        /// <summary>The ink on a paletted key: the ramp's own reading, a step brighter while the pointer is on it.</summary>
        public static Color PalettedInk(Tone tone, bool lit)
        {
            var ink = WordOn(Ramp(tone));
            return lit ? Color.Lerp(ink, Color.white, 0.35f) : ink;
        }

        /// <summary>The face colour a tone's cells are filled with (Tools/menu_pack.py measured them).</summary>
        public static Color Face(Tone tone) =>
            tone == Tone.Orange ? new Color32(230, 69, 57, 255)
            : tone == Tone.Green ? new Color32(99, 171, 63, 255)
            : new Color32(104, 111, 153, 255);

        /// <summary>The glyph's ink on a tone: the pack's resting tone, or its lit one under the pointer.</summary>
        public static Color Ink(Tone tone, bool lit)
        {
            if (lit) return tone == Tone.Green ? new Color32(255, 238, 131, 255) : new Color32(245, 255, 232, 255);
            return tone == Tone.Grey ? new Color32(223, 224, 232, 255) : new Color32(255, 194, 161, 255);
        }

        /// <summary>The ink a WORD is printed in on a tone: the grey key's resting glyph tone, and on the orange and
        /// green keys their lit tone, which is the one that reads against those faces.</summary>
        public static Color Word(Tone tone) => tone == Tone.Grey ? Ink(tone, false) : Ink(tone, true);
    }
}

File: Tools/menu_pack.py
# -*- coding: utf-8 -*-
"""THE MENU'S BUTTON PACK AND KEY CAPS (2026-09-15, the author: "KEybind için ...\\konsept art\\Classic burdaki
dosyalardaki görselleri kullan animasyonlu olduğundan 2 frame olabilir. Butonlar içinde bu dosya yolundaki butonları
kullan ...\\konsept art\\buton").

Two packs the author dropped on the Desktop, brought into Resources by this script and by nothing else:

  buton/    ten 80x96 sheets of thirty 16x16 icon buttons, five across and six down — grey, orange and green, each in
            a resting, a lit (brighter glyph) and a pressed (frame one pixel down, shadow one row thinner) state.
            Copied as they are to Resources/Menu/pack_<colour>_<state>.png; the game slices a cell at run time.
            Derived here, because the pause menu's keys are WORDS and the pack has none:
              pack_<colour>_blank.png / pack_<colour>_blank_pressed.png   one cell with its glyph painted over in
                                                                          the face colour — 9-sliced at 4, it is the
                                                                          plate a worded key stands on
              pack_glyphs.png                                             every glyph as a white mask (primary tone
                                                                          alpha 255, secondary 128) in the pack's own
                                                                          cell order, plus a seventh row of two the
                                                                          player needs and the pack lacks — the song
                                                                          before and after, built from its own play
  Classic/  key caps, 17x16 a frame, two frames a sheet (up, and pressed); a few keys wider. The Dark set is copied to
            Resources/Keys/<NAME>.png, ENTER (31 tall, L-shaped) and the ALTERNATIVE twins left behind. A key the
            pack has no cap for is drawn on EMPTY1 (one letter) or EMPTY2 (a word) with the word printed on it.

The packs keep their own palettes (GDD 16's palette-token rule bends here on the author's word, the way the flags and
the bottles keep theirs); everything the game draws around them stays in the palette.

    py -3 -X utf8 Tools/menu_pack.py        copy and derive; prints the pack's palette for MenuPack.cs
"""
import os
import shutil

from PIL import Image

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
SRC_BUTTONS = os.path.join(os.path.expanduser('~'), 'Desktop', 'konsept art', 'buton')
SRC_CAPS = os.path.join(os.path.expanduser('~'), 'Desktop', 'konsept art', 'Classic', 'Dark', 'Spritesheets')
OUT_MENU = os.path.join(ROOT, 'Assets', 'Resources', 'Menu')
OUT_KEYS = os.path.join(ROOT, 'Assets', 'Resources', 'Keys')

SHEETS = {
    'grey_normal': 'UI_grey_buttons_1.png',
    'grey_light': 'UI_grey_buttons_light_1.png',
    'grey_light_pressed': 'UI_grey_buttons_light_pressed_1.png',
    'orange_normal': 'UI_orange_buttons_3.png',
    'orange_light': 'UI_orange_buttons_light_3.png',
    'orange_light_pressed': 'UI_orange_buttons_light_pressed_3.png',
    'orange_pressed': 'UI_orange_buttons_pressed_3.png',
    'green_normal': 'UI_green_buttons_4.png',
    'green_light_pressed': 'UI_green_buttons_light_pressed_4.png',
    'green_pressed': 'UI_green_buttons_pressed_4.png',
}

# the pack's cells, read off the sheet (row-major, five a row)
ICONS = [
    'exit', 'tent', 'play', 'home', 'stats',
    'back', 'pause', 'menu', 'music', 'gamepad',
    'expand', 'sound_on', 'sound_off', 'cog', 'save',
    'restart', 'close', 'trophy', 'mail', 'heart_line',
    'check', 'trash', 'chevron_down', 'chevron_up', 'heart',
    'plus', 'minus', 'lock', 'unlock', 'info',
]
CELL = 16
INNER = (2, 3, 14, 12)      # x0, y0, x1, y1 (exclusive): where a glyph can be, inside the rim, on a resting cell

# caps that come across: every 17-wide key, and the wide ones that stay 16 tall
CAPS_SKIP = {'ENTER', 'ENTERALTERNATIVE', 'BACKSPACEALTERNATIVE', 'SHIFTALTERNATIVE', 'SPACEALTERNATIVE', 'TABALTERNATIVE'}


def cell(im, index):
    r, c = divmod(index, 5)
    return im.crop((c * CELL, r * CELL, c * CELL + CELL, r * CELL + CELL))


def anatomy(im):
    """The face, the rim highlight, the shadow, and the two glyph tones of a sheet, from its cells: the face is the
    commonest opaque colour, the shadow the colour of row 13, the highlight the colour at (2,1); the glyph tones are
    what is left inside the inner area, the primary being the one the 'expand' glyph's diagonals use."""
    counts = {}
    for i in range(30):
        for px in cell(im, i).getdata():
            if px[3]:
                counts[px] = counts.get(px, 0) + 1
    face = max(counts, key=counts.get)
    c0 = cell(im, 0)
    highlight = c0.getpixel((2, 1))
    shadow = c0.getpixel((8, 13))
    outline = c0.getpixel((8, 0))
    expand = cell(im, ICONS.index('expand'))
    primary = expand.getpixel((3, 3))          # the diagonal's first pixel
    tones = set()
    for i in range(30):
        cl = cell(im, i)
        for y in range(INNER[1], INNER[3]):
            for x in range(INNER[0], INNER[2]):
                px = cl.getpixel((x, y))
                if px[3] and px != face:
                    tones.add(px)
    secondary = [t for t in tones if t != primary]
    return dict(face=face, highlight=highlight, shadow=shadow, outline=outline, primary=primary,
                secondary=secondary[0] if len(secondary) == 1 else secondary)


def blank(im, pressed):
    """One cell with the glyph painted over in the face colour. A pressed sheet's frame sits one row lower."""
    a = anatomy(im)
    cl = cell(im, ICONS.index('info')).copy()
    x0, y0, x1, y1 = INNER
    if pressed:
        y0, y1 = y0 + 1, y1 + 1
    for y in range(y0, y1):
        for x in range(x0, x1):
            cl.putpixel((x, y), a['face'])
    return cl


def glyph_mask(im, index, pressed=False):
    a = anatomy(im)
    cl = cell(im, index)
    out = Image.new('RGBA', (CELL, CELL), (0, 0, 0, 0))
    x0, y0, x1, y1 = INNER
    if pressed:
        y0, y1 = y0 + 1, y1 + 1
    for y in range(y0, y1):
        for x in range(x0, x1):
            px = cl.getpixel((x, y))
            if px[3] and px != a['face']:
                out.putpixel((x, y), (255, 255, 255, 255 if px == a['primary'] else 128))
    return out


def derived_glyphs(play):
    """The song before and after: the pack's play triangle with a bar on the far side, and its mirror."""
    bb = play.getbbox()
    tri = play.crop(bb)
    w, h = tri.size
    nxt = Image.new('RGBA', (CELL, CELL), (0, 0, 0, 0))
    x = bb[0] - 1
    nxt.alpha_composite(tri, (x, bb[1]))
    for y in range(bb[1], bb[3]):
        for bx in (x + w + 1, x + w + 2):
            nxt.putpixel((bx, y), (255, 255, 255, 255))
    prv = nxt.transpose(Image.FLIP_LEFT_RIGHT)
    return prv, nxt


def main():
    os.makedirs(OUT_MENU, exist_ok=True)
    os.makedirs(OUT_KEYS, exist_ok=True)
    sheets = {}
    for name, fn in SHEETS.items():
        im = Image.open(os.path.join(SRC_BUTTONS, fn)).convert('RGBA')
        assert im.size == (80, 96), (fn, im.size)
        im.save(os.path.join(OUT_MENU, 'pack_%s.png' % name))
        sheets[name] = im
    for colour in ('grey', 'orange', 'green'):
        a = anatomy(sheets[colour + '_normal'])
        lit = anatomy(sheets[colour + '_light' if colour + '_light' in sheets else colour + '_light_pressed'])
        print('%-7s face %s  primary %s  secondary %s  lit primary %s  lit secondary %s  shadow %s  highlight %s'
              % (colour, a['face'][:3], a['primary'][:3], a['secondary'][:3] if isinstance(a['secondary'], tuple) else a['secondary'],
                 lit['primary'][:3], lit['secondary'][:3] if isinstance(lit['secondary'], tuple) else lit['secondary'],
                 a['shadow'][:3], a['highlight'][:3]))
        blank(sheets[colour + '_normal'], False).save(os.path.join(OUT_MENU, 'pack_%s_blank.png' % colour))
        blank(sheets[colour + '_light_pressed'], True).save(os.path.join(OUT_MENU, 'pack_%s_blank_pressed.png' % colour))

    # the glyph masks: the pack's six rows, and a seventh with the two the player lacks
    masks = Image.new('RGBA', (80, 112), (0, 0, 0, 0))
    grey = sheets['grey_normal']
    for i in range(30):
        r, c = divmod(i, 5)
        masks.alpha_composite(glyph_mask(grey, i), (c * CELL, r * CELL))
    prv, nxt = derived_glyphs(glyph_mask(grey, ICONS.index('play')))
    masks.alpha_composite(prv, (0, 6 * CELL))
    masks.alpha_composite(nxt, (1 * CELL, 6 * CELL))
    masks.save(os.path.join(OUT_MENU, 'pack_glyphs.png'))

    copied = 0
    for fn in sorted(os.listdir(SRC_CAPS)):
        name = os.path.splitext(fn)[0]
        if name in CAPS_SKIP:
            continue
        im = Image.open(os.path.join(SRC_CAPS, fn))
        assert im.size[1] == 16 and im.size[0] % 2 == 0, (fn, im.size)
        shutil.copyfile(os.path.join(SRC_CAPS, fn), os.path.join(OUT_KEYS, fn))
        copied += 1
    cap = Image.open(os.path.join(SRC_CAPS, 'B.png')).convert('RGBA')
    print('caps copied: %d   cap palette: %s' % (copied, sorted(set(px for px in cap.getdata() if px[3]), key=sum)))
    print('menu pack at', OUT_MENU)


if __name__ == '__main__':
    main()
